[1-3단계 - 방탈출 예약 관리] 리원(김지윤) 미션 제출합니다.

awrion3 opened the pull request:
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr/studylogs/4192)에 셀프 체크를 작성했나요?

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
- 페이지 출력과 함께 기본적인 crud 기능 구현과 관련하여 요구 사항을 준수했습니다. 
- 현재 시점에서는 잘못된 값과 관련된 예외에 대해서는 별도로 처리하고 있지 않습니다.
- 예약 객체와 관련하여 일급 컬렉션을 별도로 구성하지 않았습니다.

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [x] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [ ] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [ ] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

```
안녕하세요, 제이미. 이번 미션을 함께 진행하게 된 코드 리뷰이 리원입니다.
이번 미션을 진행하면서 아직 익숙하지 않은 여러 스프링 개념을 찾아볼 수 있었습니다.
그리고 현재로서는 요구 사항에 따라 알맞은 페이지가 출력되도록 구현한 단계에 있는데요, 
여기에 추가적으로 어디까지 예외 처리를 해야 할 지에 대해 생각해보고 있는 상황입니다.
시간 내어 리뷰 해주셔서 감사합니다. ☺️
```

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

awrion3 committed:
feat: add admin controller for getting main page

awrion3 committed:
feat: add read reservation

awrion3 committed:
feat: add create and delete reservation

awrion3 committed:
refactor: use dto and add reservation copy method

awrion3 committed:
refactor: use record for reservation

awrion3 committed:
docs: add simple feature list

awrion3 committed:
refactor: use require non null for reservation parameter

jamie9504 commented on src/main/java/roomescape/controller/AdminController.java:
```diff
@@ -0,0 +1,60 @@
+@Controller
+public class AdminController {
+
+    private final AtomicLong index = new AtomicLong(1);
+
+    private final List<Reservation> reservations = new ArrayList<>();
```
사이에 개행을 한 이유는 무엇인가요?

---

`@Controller`로 선언된 Controller의 역할은 무엇인가요?

awrion3 replied:
아 네, `List<Reservation>`를 분리할 지에 대해 생각하던 중 무심코 넣은 것 같습니다, 해당 개행은 제거하는 방향으로 리팩토링 했습니다.
그리고 `@Controller`로 선언한 클래스의 경우, 스프링 빈으로 등록되어 웹 클라이언트로부터 들어오는 HTTP 요청을 처리하고 응댭을 반환하는 역할을 하게 됩니다.
이를 고려해 Controller에 있던 `Reservation`과 관련된 조회, 추가, 삭제의 비즈니스 로직을 분리하고자, `Reservations` 클래스를 도입해 리팩토링 해봤습니다.

jamie9504 commented on src/main/java/roomescape/controller/AdminController.java:
```diff
@@ -0,0 +1,60 @@
+        return "/admin/reservation-legacy";
+    }
+
+    @GetMapping("/reservations")
+    @ResponseBody
+    public ResponseEntity<List<ReservationDto>> readReservations() {
+        List<ReservationDto> reservationDtos = reservations.stream().map(ReservationDto::of).toList();
+        return ResponseEntity.ok().body(reservationDtos);
+    }
+
+    @PostMapping("/reservations")
+    @ResponseBody
+    public ResponseEntity<Reservation> createReservation(@RequestBody Reservation reservation) {
+        Reservation newReservation = reservation.copyWithId(index.getAndIncrement());
+        reservations.add(newReservation);
+        return ResponseEntity.ok().body(newReservation);
+    }
+
+    @DeleteMapping("/reservations/{id}")
+    public ResponseEntity<Void> deleteReservation(@PathVariable("id") Long id) {
+        Reservation reservation = reservations.stream()
+                .filter(reserve -> Objects.equals(reserve.id(), id))
+                .findFirst()
+                .orElseThrow(RuntimeException::new);
+
+        reservations.remove(reservation);
+        return ResponseEntity.ok().build();
```
Controller명이 AdminController인데,
해당 메서드들이 Admin이라는 정보와 연관이 되어 있을까요?

awrion3 replied:
음, 그렇네요. AdminController라는 이름은 의미 측면에서 다소 명료하지 않은 것 같네요!
리팩토링 과정에서는 어드민 페이지 출력의 `AdminDisplayController`와 예약 내역 관리의 `AdminReservationsController`로 나누어 재정의 해보았습니다.

jamie9504 commented on src/main/java/roomescape/dto/ReservationDto.java:
```diff
@@ -0,0 +1,16 @@
+public record ReservationDto(long id, String name, LocalDate date, String time) {
+
+    public static ReservationDto of(Reservation reservation) {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("HH:mm");
```
매번 formatter를 생성하고 있군요.

awrion3 replied:
앗 네, 해당 부분은 상수로 바꾸어 리팩토링 했습니다!

jamie9504 commented on src/main/java/roomescape/controller/AdminController.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @GetMapping("/reservations")
+    @ResponseBody
```
`@RestController`에 대해서도 학습해보세요.

awrion3 replied:
네, 우선 `@Controller`의 경우 Spring MVC의 컨트롤러로서 클라이언트의 요청을 받아 주로 뷰를 반환하게 됩니다.
그리고 응답과 관련하여 본문의 데이터를 반환하고자 할 때는 `@ResponseBody`를 사용하며, 이렇게 반환된 객체는 json 데이터로 직렬화 됩니다.
한편 `@RestController`는 `@Controller`에 `@ResponseBody`를 더한 것이라는 점에서, 마찬가지로 객체를 반환할 때 사용된다고 이해했습니다.
또한 json 형식의 응답이 기본이 되므로, 뷰 템플릿 없이 RESTful API 중심의 구조를 만들 때 좀 더 적합하다고 생각했습니다.

jamie9504 commented on src/main/java/roomescape/controller/AdminController.java:
```diff
@@ -0,0 +1,60 @@
+
+    @GetMapping("/reservations")
+    @ResponseBody
+    public ResponseEntity<List<ReservationDto>> readReservations() {
```
ResponseEntity로 감싸는 것과 감싸지 않는 것은 어떠한 차이가 있나요?

awrion3 replied:
네, `@ResponseBody`를 사용해서 객체를 반환하면, HTTP 응답과 관련해 body에 대해서만 설정할 수 있습니다.
반면 `ResponseEntity`로 감싸서 반환하는 경우, 응답의 본문 외에도 상태 코드 및 헤더에 대해서도 유연하게 설정할 수 있게 됩니다.
따라서 요청 처리 결과에 따라 `200 OK` 외 status를 설정하거나, 응답에 커스텀 헤더를 붙일 때는 `ResponseEntity`를 쓰는 것이 편리하다고 생각했습니다.

jamie9504 commented on src/main/java/roomescape/controller/AdminController.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @DeleteMapping("/reservations/{id}")
+    public ResponseEntity<Void> deleteReservation(@PathVariable("id") Long id) {
```
이 친구는 `@ResponseBody`를 선언하지 않은 이유가 무엇인가요?

awrion3 replied:
오 네, 다시 살펴보니 http 응답 생성에 있어서 `ResponseEntity`를 사용하면, `@ResponseStatus`나 `@ResponseBody`를 굳이 사용하지 않아도 되었네요. 
Create, Read 관련 메서드와 관련해서도 해당 어노테이션을 제외하는 방향으로 리팩토링 했습니다.

jamie9504 replied:
잘 알아보셨네요.

각 방법의 차이는 있지만, 리원이 생각하는 장점이 큰 방법을 선택하면 좋은데요.
ResponseEntity와 @ResponseStatus나 @ResponseBody를 함께 사용하는 것과 하나만 사용하는 것은 어떠한 차이가 있을까요? 기능적인 부분과 개발자의 직관적인 부분 및 리팩터링의 영역을 함께 고민해보세요!

jamie9504 commented on src/main/java/roomescape/controller/AdminController.java:
```diff
@@ -0,0 +1,60 @@
+
+    @PostMapping("/reservations")
+    @ResponseBody
+    public ResponseEntity<Reservation> createReservation(@RequestBody Reservation reservation) {
```
컨트롤러 내에서 reservation을 사용하는 곳과 reservationDto를 사용하는 곳이 있는데 어떤 기준으로 사용여부를 결정하고 있나요?

awrion3 replied:
음 네, 예약 내역 조회의 경우 시간 출력 포맷과 관련해서 `HH:mm`을 적용하기 위해 dto를 사용하게 되었습니다. 
그리고 예약 내역 추가의 경우 결국 반환된 객체가 json 데이터로 저장된다는 점에서 굳이 사용하지 않아도 된다고 판단했습니다.

jamie9504 commented on src/main/java/roomescape/controller/AdminController.java:
```diff
@@ -0,0 +1,60 @@
+    @PostMapping("/reservations")
+    @ResponseBody
+    public ResponseEntity<Reservation> createReservation(@RequestBody Reservation reservation) {
+        Reservation newReservation = reservation.copyWithId(index.getAndIncrement());
```
copy를 하여 사용하면 어떠한 이점이 있나요?

awrion3 replied:
음, increment한 id 값을 반영한 `Reservation` 객체를 생성함에 있어, 생성자 대신 copy 메서드를 도입해보았는데요. 
이를 통해 `@RequestBody`로 들어온 기존 객체의 id 값을 수정하지 않고, 객체를 복사해 새롭게 생성함으로써 불변성을 유지하고자 했습니다. 
다만 이러한 변환 과정을 dto인`ReservationRequest`에 위임해도 괜찮지 않을까 생각해서, `toReservation` 메서드를 추가하는 방향으로 리팩토링 했습니다.

jamie9504 commented on src/main/java/roomescape/controller/AdminController.java:
```diff
@@ -0,0 +1,60 @@
+        Reservation reservation = reservations.stream()
+                .filter(reserve -> Objects.equals(reserve.id(), id))
+                .findFirst()
+                .orElseThrow(RuntimeException::new);
```
RuntimeException이 적절한가요?

awrion3 replied:
아 그렇네요, 하위 클래스인 `IllegalArgumentException`을 사용하여 부적절한 id 값과 관련해 좀 더 구체적인 예외를 던질 수 있도록 수정했습니다!

jamie9504 reviewed (changes requested):
안녕하세요, 리원. 리뷰어 제이미입니다.

만나서 반갑습니다.

코멘트 남겨두었습니다.

awrion3 committed:
docs: update feature list

awrion3 committed:
refactor(Reservation): remove copy method and use exception header

- add null value exception test

awrion3 committed:
feat(Reservations): add find, save, delete by id methods

- add duplication check and test cases

awrion3 committed:
feat: add request and response dto for conversion

awrion3 committed:
refactor: split controllers and test cases

- add controllers for displaying pages and handling reservations
- add line break for application class

awrion3 committed:
refactor: use constant variable name for formatter

awrion3 committed:
refactor: use simplified response entity method

jamie9504 commented on src/main/java/roomescape/controller/AdminDisplayController.java:
```diff
@@ -0,0 +1,18 @@
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.GetMapping;
+
+@Controller
```
`@RequestMapping`에 대해서도 학습해보세요.

awrion3 replied:
아 네, `@RequestMapping`을 사용 시, URL 경로 value와 RequestMethod.GET와 같은 method를 설정하게 됩니다.
또한 이러한 방식 대신 `@GetMapping`, `@PostMapping` 등으로도 간단하게 사용 가능합니다.
그리고 `@RequestMapping`을 클래스 레벨에서 사용하면, 각 HTTP 메서드에서 공통되는 URL 경로를 지정할 수 있음을 알게 되었습니다.

jamie9504 replied:
새로운 개념을 익힐 때, 대체할 수 있는 것과 비슷한 것들을 함께 익힌다면 조금 더 넓은 학습이 가능합니다.

jamie9504 commented on src/main/java/roomescape/controller/AdminDisplayController.java:
```diff
@@ -0,0 +1,18 @@
+
+    @GetMapping("/admin")
+    public String displayAdminMain() {
+        return "/admin/index";
```
`/admin`이라는 path가 return에서 중복되고 있네요.

awrion3 replied:
네, 해당 부분은 이번에 학습한 `@RequestMapping`을 사용해서 중복되는 경로 부분을 제거하는 방향으로 수정했습니다!

jamie9504 commented on src/main/java/roomescape/model/Reservation.java:
```diff
@@ -0,0 +1,16 @@
+
+public record Reservation(long id, String name, LocalDate date, LocalTime time) {
+
+    private static final String INVALID_INPUT = "[EXCEPTION] 널 값은 입력될 수 없습니다.";
```
상수를 분리하는 경우, 정확하게 의미를 담아주어야해요. 해당 상수는 예외 메시지를 의미하는 거겠죠?

awrion3 replied:
네, 예외 메시지 관련 맞습니다! 현재 상수명보다 좀 더 의미를 잘 드러내는 방향으로 리팩토링 해보겠습니다.

jamie9504 commented on src/main/java/roomescape/model/Reservations.java:
```diff
@@ -0,0 +1,46 @@
+
+    private final List<Reservation> reservations;
+
+    public Reservations() {
+        this.reservations = new ArrayList<>();
+    }
```
생성자를 구현할 필요가 있을까요?

awrion3 replied:
음, 이러한 경우 필드에서 초기화 해도 괜찮겠다는 생각이 드네요. 다만 리팩토링 과정에서는 확장성을 고려해 생성자를 활용한 DI 방식으로 수정해보게 되었습니다.

jamie9504 commented on src/test/java/roomescape/controller/AdminReservationsControllerTest.java:
```diff
@@ -0,0 +1,57 @@
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.annotation.DirtiesContext;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
+@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
```
이번 미션을 진행하면서, 1레벨에서 사용하지 않았던 애너테이션, 클래스 등을 마주했을 텐데요.
이 부분에 대해서 현재 사용하는 방법, 그리고 다른 방식으로 사용하는 방법들을 어느정도 학습하셨나요?

[숙제] 위의 내용에 대해 현재 학습한 것 보다 1 depth 더 깊거나 넓게 학습해보고, 이 부분을 정리해주세요.

awrion3 replied:
아 네, 우선 `@SpringBootTest`에서는 webEnvironment 속성을 통해 테스트 실행 시 웹 환경의 종류를 선택할 수 있습니다.
예를 들어, `DEFINED_PORT`는 application.properties에서 별도로 설정하지 않은 한 8080 포트에서 실제 서버를 구동합니다.
그리고 `RANDOM_PORT`는 충돌 방지를 위해 랜덤 포트를 사용합니다.
또한 기본값인 `MOCK`은 실제 서버 없이 가짜 웹 환경을 제공하며, `NONE`은 웹 환경 없이 일반 애플리케이션 컨텍스트만 로드합니다.

그리고 `@DirtiesContext`의 경우, 테스트 중 애플리케이션 컨텍스트가 변경되었음을 스프링에 알리는 어노테이션입니다.
이와 관련하여 클래스 레벨 설정과 관련해서는,
`BEFORE_EACH_TEST_METHOD`는 테스트 클래스의 각 테스트 메서드 실행 전마다 컨텍스트 초기화를 수행합니다.
`AFTER_EACH_TEST_METHOD`는 테스트 클래스의 각 테스트 메서드 실행 후마다 컨텍스트 초기화를 수행합니다.
`BEFORE_CLASS`는 테스트 클래스 실행 전에 컨텍스트 초기화를 수행합니다.
`AFTER_CLASS`는 테스트 클래스 실행 후에 컨텍스트 초기화를 수행합니다.
또한 메서드 레벨 설정과 관련해서는,
`BEFORE_METHOD`는 해당 테스트 메서드 실행 전에 컨텍스트 초기화를 수행합니다.
`AFTER_METHOD`는 해당 테스트 메서드 실행 후에 컨텍스트 초기화를 수행합니다.
이러한 설정을 통해 스프링은 해당 컨텍스트를 닫고 새로 생성하도록 지시받게 되며, 다른 테스트에 영향을 주지 않도록 테스트 간 격리를 보장하는 데 사용될 수 있습니다.

jamie9504 replied:
좋습니다. 이번 레벨에서 새로운 것을 접하고 사용할 때, 이와 같이 학습하는 습관을 들이면
레벨이 끝났을 때 리원이 어떤 것을 왜 사용하는지 설명하기에도 용이하고, 구현하는 다른 방식을 선택할 수도 있을 거에요.

jamie9504 commented on src/main/java/roomescape/controller/AdminReservationsController.java:
```diff
@@ -0,0 +1,40 @@
+import roomescape.model.Reservation;
+import roomescape.model.Reservations;
+
+@Controller
```
HttpStatus는 OK만 내려주고 있는데, 굳이 ResponseEntity로 내려줄 필요가 있을까요?

awrion3 replied:
음 네, 현 시점에서는 HTTP status와 관련해서 ok로만 설정하고 있는데요. 
ResponseEntity를 사용하면 추후 필요 시 다른 status로 설정을 변경함에 있어서 유연성을 제공해줄 수 있다고 생각해서 적용하게 되었습니다.

jamie9504 replied:
추후 다른 Status Code로 바꿀 때, ResponseEntity를 사용하는 방식으로 변경하는 비용이 클까요?
변경을 할 수도 있고, 하지 않을 수도 있는 부분인데 이 부분을 미리 고려한 이유는 무엇인가요?

awrion3 replied:
아 네, 변경하지 않는 경우 이를 수정하는 데 드는 비용이 더 클 수도 있다는 생각이 드네요.
그리고 ResponseEntity를 그럼에도 불구하고 사용한 이유는, 현재 시점에서는 ok()로 모두 상태 코드를 지정하고 있는데요.
`noContent()`와 같은 상태 코드를 2단계에서 도입할지 고민하고 있어서 이렇게 사용해보게 된 것 같습니다.

jamie9504 commented on src/main/java/roomescape/controller/AdminReservationsController.java:
```diff
@@ -0,0 +1,40 @@
+@Controller
+public class AdminReservationsController {
+
+    private final AtomicLong id = new AtomicLong(1);
+    private final Reservations reservations = new Reservations();
```
ID를 관리하는 것이 Controller의 책임일까요?

awrion3 replied:
음 네, 말씀대로 Controller가 가질 책임은 아니라고 생각되네요. 해당 로직은 `Id` 객체를 새롭게 도입하는 것으로 변경했습니다!

jamie9504 commented on src/main/java/roomescape/controller/AdminReservationsController.java:
```diff
@@ -0,0 +1,40 @@
+
+    @PostMapping("/reservations")
+    public ResponseEntity<Reservation> createReservation(@RequestBody ReservationRequest reservationRequest) {
+        Reservation reservation = reservations.saveReservation(reservationRequest.toReservation(id.getAndIncrement()));
```
소괄호가 계속 중첩되면, 코드를 좌->우로 읽지 못하게 되어 가독성이 저해됩니다.

awrion3 replied:
아 네, 해당 부분은 `newReservation`과 `savedReservation`으로 좀 더 분리해서 리팩토링 했습니다!

jamie9504 reviewed (changes requested):
안녕하세요, 리원. 리뷰어 제이미입니다.

코멘트 남겨두었습니다.

다음 리뷰 요청시 특이사항이 없다면, Merge 예정입니다.

awrion3 committed:
refactor: use id class for atomic long

- use constant for controller test
- use request mapping for controller

awrion3 committed:
refactor: use di and final for reservations

- rename constant for exception message
- apply synchronized keyword for reservations

jamie9504 commented on src/main/java/roomescape/controller/AdminReservationsController.java:
```diff
@@ -0,0 +1,43 @@
+    @GetMapping
+    public ResponseEntity<List<ReservationResponse>> readReservations() {
+        List<ReservationResponse> responses = ReservationResponse.toResponses(reservations.findReservations());
+        return ResponseEntity.ok(responses);
```
ResponseEntity를 사용하는 여러 방법에 대해서도 학습해보세요.

jamie9504 commented on src/main/java/roomescape/dto/ReservationResponse.java:
```diff
@@ -0,0 +1,24 @@
+
+public record ReservationResponse(long id, String name, LocalDate date, String time) {
+
+    private static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("HH:mm");
```
조금 더 직관적인 이름을 지어주세요.

jamie9504 commented on README.md:
```diff
@@ -0,0 +1,26 @@
+
+---
+
+### 기능 목록: 1 ~ 3단계
```
단계별 요구사항을 정리한 건 맞으나, 기능 목록을 정리했다고 볼 수 있을까요?

jamie9504 commented on src/main/java/roomescape/model/Reservations.java:
```diff
@@ -0,0 +1,46 @@
+public class Reservations {
+
+    private static final String IS_DUPLICATE_TIME = "[EXCEPTION] 해당 시간대는 이미 예약되어 있습니다.";
+    private static final String IS_INVALID_ID = "[EXCEPTION] 해당 아이디는 존재하지 않습니다.";
```
직관적으로 상수명을 지어달라는 의미는 예를 들자면

IS_INVALID_ID라는 이름만 보았을 때는 boolean 타입에 무언가를 반환할 것이라 생각할 수도 있습니다.

해당 상수명만 보았을 때, 예외 메시지라는 의미가 담겨있으면 좋겠어요.

jamie9504 commented on src/main/java/roomescape/model/Reservation.java:
```diff
@@ -0,0 +1,16 @@
+
+public record Reservation(long id, String name, LocalDate date, LocalTime time) {
+
+    private static final String IS_NULL_VALUE = "[EXCEPTION] 널 값은 입력될 수 없습니다.";
```
앞에 `[EXCEPTION] `을 붙이는 이유는 무엇인가요?

jamie9504 commented on src/main/java/roomescape/model/Reservations.java:
```diff
@@ -0,0 +1,46 @@
+        this.reservations = new ArrayList<>(reservations);
+    }
+
+    public List<Reservation> findReservations() {
+        return reservations.stream().toList();
+    }
+
+    public synchronized Reservation saveReservation(final Reservation reservation) {
+        checkDuplicates(reservation);
+        reservations.add(reservation);
+        return reservation;
+    }
+
+    public synchronized void deleteReservationById(final Long id) {
+        Reservation reservation = reservations.stream()
+                .filter(reserve -> Objects.equals(reserve.id(), id))
+                .findFirst()
+                .orElseThrow(() -> new IllegalArgumentException(IS_INVALID_ID));
+        reservations.remove(reservation);
```
개행을 해주는 방식이 전반적으로 통일이 되어있지 않은데, 코드 스타일도 맞춰볼까요?

jamie9504 reviewed (commented):
안녕하세요, 리원. 리뷰어 제이미입니다.

코멘트 남겨두었습니다.

이번 단계 요구사항은 모두 충족한 것 같아 이만 Approve & Merge합니다.

머지한 PR을 재확인하지 않기에 코멘트에 대한 답변은 다음 단계 PR 본문에 작성해주세요.
